xwiki-platform-core/xwiki-platform-search/xwiki-platform-search-solr/xwiki-platform-search-solr-api/src/main/java/org/xwiki/search/solr/AbstractSolrCoreInitializer.java

```diff
@@ -890,28 +877,10 @@ protected void setField(String name, String type, boolean dynamic, Object... att
      */
     protected void setField(Map<String, Object> fieldAttributes, boolean dynamic) throws SolrException
     {
-        String name = (String) fieldAttributes.get(SOLR_FIELD_NAME);
 
-        try {
-            if (getFields(false).containsKey(name)) {
-                if (dynamic) {
-                    new SchemaRequest.ReplaceDynamicField(fieldAttributes).process(this.client);
-                } else {
-                    new SchemaRequest.ReplaceField(fieldAttributes).process(this.client);
-                }
-            } else {
-                if (dynamic) {
-                    new SchemaRequest.AddDynamicField(fieldAttributes).process(this.client);
-                } else {
-                    new SchemaRequest.AddField(fieldAttributes).process(this.client);
-                }
-            }
-
-            // Add it to the cache
-            this.fields.put(name, fieldAttributes);
-        } catch (Exception e) {
-            throw new SolrException("Failed to set a field in the Solr core", e);
-        }
+        this.solrSchemaUtils.setField(getCoreName(), fieldAttributes, dynamic);
+        // Ensure to keep internal fields updated
+        getFields(false);
```

Comment:
This call seems useless. The new field is already added to the cache by #setField.

Reply:
Well not really anymore: `SolrSchemaUtils#setField` does add the new field to the `SolrSchemaUtils.SolrSchema#fields` but in AbstractSolrCoreInitializer, `fields` is a copy of  `SolrSchemaUtils.SolrSchema#fields` and not a reference to it. See: https://github.com/xwiki/xwiki-platform/pull/2066/files#diff-4b1b8c4ad281f72b8ea3f2415bf4111077639746b3ab520f7ba134958173c4b0R101

I hesitated between having a pointer or a copy to it, but it felt safer to return a copy to avoid any concurrency issue: by documenting that I want also to prevent any inconsistency in case someone calls commit, which empty the field map.

Reply:
> avoid any concurrency issue

Well this copy is really error-prone and, to me, all it takes to avoid concurrency issues with the fields is to store a ConcurrentHAshMap instead of a HashMap. You don't even need to have a `fields` field on `AbstractSolrCoreInitializer` side.

Reply:
> You don't even need to have a fields field on AbstractSolrCoreInitializer side.

problem is that it's protected right now, so it's an API. 

> all it takes to avoid concurrency issues with the fields is to store a ConcurrentHashMap instead of a HashMap

yes I thought about that (was my first implementation actually), but in case of a call to `SolrSchemaUtils#commit` it means `AbstractSolrCoreInitializer#fields` will keep pointing an outdated map. Now maybe it's less error prone.

Reply:
So I changed my mind and I now use a ConcurrentHashMap. Not perfect, but I'm planning to improve all this in the future.